File: 2018_05_01/SocketLearn1/SocketLearn1/RootVC/RootViewController.m
//
//  RootViewController.m
//  SocketLearn1
//
//  Created by xiaoyulong on 2018/5/31.
//  Copyright © 2018年 xiaoyulong. All rights reserved.
//

#import "RootViewController.h"

@interface RootViewController ()

@end

@implementation RootViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


// MARK: - 将文件 AsyncSocket 拖进来，
// MARK: —>TARGETS  —> Build Phases —> Compile Sources 中
// MARK: -fno-objc-arc
- (void)demo1
{
    
}


// MARK: - 什么是 Socket？
- (void)demo2
{
    
    /*
     我们已经知道网络中的进程是通过socket来通信的，那什么是socket呢？
     socket起源于Unix，而Unix/Linux基本哲学之一就是“一切皆文件”，都可以用，“打开open –> 读写write/read –> 关闭close”模式来操作。
     我的理解就是Socket就是该模式的一个实现，socket即是一种特殊的文件，一些socket函数就是对其进行的操作（读/写IO、打开、关闭）
     这些函数我们在后面进行介绍。我们在传输数据时，可以只使用（传输层）TCP/IP协议，
     但是那样的话，如果没有应用层，便无法识别数据内容，如果想要使传输的数据有意义，
     则必须使用到应用层协议，应用层协议有很多，比如HTTP、FTP、TELNET等，
     也可以自己定义应用层协议。
     WEB使用HTTP协议作应用层协议，以封装HTTP文本信息，
     然后使用TCP/IP做传输层协议将它发到网络上。
     
     
     1)Socket是一个针对TCP和UDP编程的接口，你可以借助它建立TCP连接等等。而TCP和UDP协议属于传输层。
     而http是个应用层的协议，它实际上也建立在TCP协议之上。
     (HTTP是轿车，提供了封装或者显示数据的具体形式；Socket是发动机，提供了网络通信的能力。)
     2）Socket是对TCP/IP协议的封装，Socket本身并不是协议，而是一个调用接口（API），通过Socket，我们才能使用TCP/IP协议。
     Socket的出现只是使得程序员更方便地使用TCP/IP协议栈而已，
     是对TCP/IP协议的抽象，从而形成了我们知道的一些最基本的函数接口。
     */
    
    
    
    
    /**
     socket连接和http连接的区别
     简单说，你浏览的网页（网址以http://开头)都是http协议传输到你的浏览器的,
     而http是基于socket之上的。socket是一套完成tcp，udp协议的接口。
     
     HTTP协议：简单对象访问协议，对应于应用层  ，HTTP协议是基于TCP连接的
     tcp协议：    对应于传输层
     ip协议：     对应于网络层
     
     TCP/IP是传输层协议，主要解决数据如何在网络中传输；而HTTP是应用层协议，主要解决如何包装数据。
     
     Socket是对TCP/IP协议的封装，Socket本身并不是协议，而是一个调用接口（API），通过Socket，我们才能使用TCP/IP协议。
     http连接：http连接就是所谓的短连接，即客户端向服务器端发送一次请求，服务器端响应后连接即会断掉；
     
     socket连接：socket连接就是所谓的长连接，理论上客户端和服务器端一旦建立起连接将不会主动断掉；
     但是由于各种环境因素可能会是连接断开，比如说：服务器端或客户端主机down了，
     网络故障，或者两者之间长时间没有数据传输，
     网络防火墙可能会断开该连接以释放网络资源。
     所以当一个socket连接中没有数据的传输，那么为了维持连接需要发送心跳消息~~
     具体心跳消息格式是开发者自己定义的
     */
}


// MARK: - 利用Socket建立网络连接的步骤
- (void)demo3
{
    /**
     利用Socket建立网络连接的步骤：
     
     建立Socket连接至少需要一对套接字，其中一个运行于客户端，称为ClientSocket；
     另一个运行于服务端，称为ServerSocket；
     
     套接字的连接过程分为三个步骤：
     1.服务器监听：服务器端套接字并不定位具体的客户端套接字，而是处于等待连接的状态，
     实时监控网络状态，等待客户端的连接请求。
     
     2.客户端请求：指客户端的套接字提出连接请求，要连接的目标是服务器端的套接字。
     为此，客户端的套接字必须首先描述它要连接的服务器的套接字，
     指出服务器端套接字的地址和端口号，然后就向服务器端套接字提出连接请求。
     
     3.连接确认：当服务器端套接字监听到或者说接收到客户端套接字的连接请求时，
     就响应客户端套接字的请求，建立一个新的线程，
     把服务器端套接字的描述发给客户端，
     一旦客户端确认了此描述，双方就正式建立连接。
     而服务器端套接字继续处于监听状态，继续接收其他客户端套接字的连接请求。
     */
    
    
    /*
     IM通讯协议（基于TCP/IP和UDP）
     1  A登录 IM服务器纪录A的TCP／UDP端口 ＋ A的ip 地址  A在线
     2  向A的好友列表（存储在IM上）发送A的端口＋IP地址
     3  将A的好友列表 端口＋ip＋在线状态等信息传回登录A的PC 机
     
     IM服务器还提供消息中转服务
     网络好  不中转  点对点通讯
     网络不好   中转（消息先到服务器再到A或者B）
     
     IM通讯方式 xmpp
     1 在线直接通讯   （点对点）
     2 在线代理通讯   （网络情况不好  防火墙 消息先到服务器再转）
     3 离线代理通讯  （服务起保存消息，等你上线  再把消息给你）
     4 扩展方式通讯   （发到手机 ，email,传真）
     */
}


@end

File: 2017_01_01/RSA_DES_AES_MD5/RSA_DES_AES_MD5/RootVC/RootViewController.swift
//
//  RootViewController.swift
//  RSA_DES_AES_MD5
//
//  Created by xiaoyulong on 17/1/17.
//  Copyright © 2017年 xiaoyulong. All rights reserved.
//

import UIKit

class RootViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        /**
         
         网络安全之数据加密（DES、AES、RSA、MD5）
         http://www.jianshu.com/p/98569e81cc0b
         https://github.com/watermin
         
         
         iOS，一行代码进行RSA、DES 、AES加密、解密及MD5加密
         http://www.jianshu.com/p/8fd8306a95d0
         https://github.com/Flying-Einstein/Encryptions
         
         */
     
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    // MARK: - MD5
    // MARK: 是一种校验方式，用于保证文件的正确性，防止被植入木马或病毒。
    // MARK: MD5功能：输入任意长度的信息，经过处理，输出为128位的信息（数字指纹）；不同的输入得到的不同的结果（唯一性）；根据128位的输出结果不可能反推出输入的信息（不可逆）。
    func demo4() -> () {
        
        /**
         
         MD5 即 Message-Digest Algorithm 5（信息-摘要算法5），
         用于确保信息传输完整一致。是计算机广泛使用的杂凑算法之一（又译摘要算法、哈希算法），
         主流编程语言普遍已有MD5实现。
         将数据（如汉字）运算为另一固定长度值，是杂凑算法的基础原理，MD5的前身有MD2、MD3和MD4。
         
         MD5算法具有以下特点：
         1、压缩性：任意长度的数据，算出的MD5值长度都是固定的。
         2、容易计算：从原数据计算出MD5值很容易。
         3、抗修改性：对原数据进行任何改动，哪怕只修改1个字节，所得到的MD5值都有很大区别。
         4、强抗碰撞：已知原数据和其MD5值，想找到一个具有相同MD5值的数据（即伪造数据）是非常困难的。
         
         MD5的作用是让大容量信息在用数字签名软件签署私人密钥前被"压缩"成一种保密的格式（就是把一个任意长度的字节串变换成一定长的十六进制数字串）。除了MD5以外，其中比较有名的还有sha-1、RIPEMD以及Haval等。
         */
        
        
        /**
         MD5（单向散列算法）的全称是Message-Digest Algorithm 5（信息-摘要算法），经MD2、MD3和MD4发展而来。MD5算法的使用不需要支付任何版权费用。
         MD5功能：输入任意长度的信息，经过处理，输出为128位的信息（数字指纹）；不同的输入得到的不同的结果（唯一性）；根据128位的输出结果不可能反推出输入的信息（不可逆）。
         MD5用途：
         a，防止被篡改：1）比如发送一个电子文档，发送前，我先得到MD5的输出结果a。然后在对方收到电子文档后，对方也得到一个MD5的输出结果b。如果a与b一样就代表中途未被篡改。2）比如我提供文件下载，为了防止不法分子在安装程序中添加木马，我可以在网站上公布由安装文件得到的MD5输出结果。3）SVN在检测文件是否在CheckOut后被修改过，也是用到了MD5.
         b，防止直接看到明文：现在很多网站在数据库存储用户的密码的时候都是存储用户密码的MD5值。这样就算不法分子得到数据库的用户密码的MD5值，也无法知道用户的密码。当用户登录的时候，系统把用户输入的密码计算成MD5值，然后再去和保存在文件系统中的MD5值进行比较，进而确定输入的密码是否正确。通过这样的步骤，系统在并不知道用户密码的明码的情况下就可以确定用户登录系统的合法性。
         无论在Android上还是pc上用Java实现MD5都比较容易，因为java已经把它做到了java.security.MessageDigest里。下面是一个MD5Util.java类的使用
         */
        
    }
    
    // MARK: - AES
    // MARK: 对称加密算法
    func demo3() -> () {
        
        /**
         
         高级加密标准（英语：Advanced Encryption Standard，缩写：AES），
         在密码学中又称Rijndael加密法，
         是美国联邦政府采用的一种区块加密标准。
         这个标准用来替代原先的DES，已经被多方分析且广为全世界所使用。
         经过五年的甄选流程，高级加密标准由美国国家标准与技术研究院（NIST）于2001年11月26日发布于FIPS PUB 197，并在2002年5月26日成为有效的标准。
         2006年，高级加密标准已然成为对称密钥加密中最流行的算法之一。
         
         */
        
    }
    
    
    // MARK: - DES
    // MARK: 对称加密算法，指加密和解密使用相同密钥的加密算法
    func demo2() -> () {
        
        /**
         des对称加密，是一种对称加密算法。
         
         在对称加密算法中，数据发信方将明文（原始数据）和加密密钥（mi yao）一起经过特殊加密算法处理后，使其变成复杂的加密密文发送出去。
         收信方收到密文后，若想解读原文，则需要使用加密用过的密钥及相同算法的逆算法对密文进行解密，才能使其恢复成可读明文。
         在对称加密算法中，使用的密钥只有一个，发收信双方都使用这个密钥对数据进行加密和解密，这就要求解密方事先必须知道加密密钥。
         */
        
        
        /**
         对称加密算法的特点是算法公开、计算量小、加密速度快、加密效率高。
         不足之处是，交易双方都使用同样钥匙，安全性得不到保证。
         此外，每对用户每次使用对称加密算法时，都需要使用其他人不知道的惟一钥匙，
         这会使得发收信双方所拥有的钥匙数量呈几何级数增长，密钥管理成为用户的负担。
         对称加密算法在分布式网络系统上使用较为困难，主要是因为密钥管理困难，使用成本较高。
         而与公开密钥加密算法比起来，对称加密算法能够提供加密和认证却缺乏了签名功能，
         使得使用范围有所缩小。在计算机专网系统中广泛使用的对称加密算法有DES和IDEA等。
         美国国家标准局倡导的AES即将作为新标准取代DES。
         
         
         基于“对称密钥”的加密算法主要有DES、TripleDES、RC2、RC4、RC5和Blowfish等。
         */
    }
    
    
    // MARK: - RSA
    // MARK: 非对称加密算法，公钥和私钥不一致
    func demo1() -> () {
        
        /**
         RSA公钥加密算法是1977年由
         罗纳德·李维斯特（Ron Rivest）、
         阿迪·萨莫尔（Adi Shamir）和
         伦纳德·阿德曼（Leonard Adleman）一起提出的。
         
         1987年首次公布，当时他们三人都在麻省理工学院工作。
         RSA就是他们三人姓氏开头字母拼在一起组成的。
         
         RSA是目前最有影响力的公钥加密算法，它能够抵抗到目前为止已知的绝大多数密码攻击，
         已被ISO推荐为公钥数据加密标准。
         */
        
        /**
         
         由于进行的都是大数计算，使得RSA最快的情况也比DES慢上好几倍，
         无论是软件还是硬件实现。速度一直是RSA的缺陷。
         一般来说只用于少量数据加密。RSA的速度比对应同样安全级别的对称密码算法要慢1000倍左右。
         
         比起DES和其它对称算法来说，RSA要慢得多。
         实际上Bob一般使用一种对称算法来加密他的信息，然后用RSA来加密他的比较短的对称密码，
         然后将用RSA加密的对称密码和用对称算法加密的消息送给Alice。
         
         这样一来对随机数的要求就更高了，尤其对产生对称密码的要求非常高，
         因为否则的话可以越过RSA来直接攻击对称密码。
         */
        
        
        
        
    }
    

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

}
